flow/aggregates.py

```diff
@@ -0,0 +1,463 @@
+        """
+        if isinstance(key, str):
+            if default is None:
+                def keyfunction(job):
+                    return job.sp[key]
+            else:
+                def keyfunction(job):
+                    return job.sp.get(key, default)
+        elif isinstance(key, Iterable):
+            keys = list(key)
+            if default is None:
+                def keyfunction(job):
+                    return [job.sp[key] for key in keys]
+            else:
+                if isinstance(default, Iterable):
+                    if len(default) != len(keys):
+                        raise ValueError("Expected length of default argument is "
+                                         f"{len(keys)}, got {len(default)}.")
+                else:
+                    raise TypeError("Invalid default argument. Expected Iterable, "
+                                    f"got {type(default)}")
+
+                def keyfunction(job):
+                    return [job.sp.get(key, default_value)
+                            for key, default_value in zip(keys, default)]
+        elif callable(key):
+            keyfunction = key
+        else:
+            raise TypeError("Invalid key argument. Expected either str, Iterable "
+                            f"or a callable, got {type(key)}")
+
+        def aggregator(jobs):
+            for key, group in groupby(sorted(jobs, key=keyfunction), key=keyfunction):
+                yield group
+
+        return cls(aggregator, sort_by, reverse_order, select, key=key, default=default)
+
+    def __eq__(self, other):
+        if type(self) != type(other):
+            return False
+        elif(
+            self._sort_by != other._sort_by or
+            self._reverse_order != other._reverse_order or
+            self._kwargs != other._kwargs
+        ):
+            return False
+        elif(
+            self._select == other._select and
+            self._aggregator == other._aggregator
+        ):
+            return True
+
+        # Get unique id for _select attribute
+        self_select = self._get_unique_function_id(self._select)
+        other_select = self._get_unique_function_id(other._select)
+        # Get unique id for _aggregator attribute
+        self_aggregator = self._get_unique_function_id(self._aggregator)
+        other_aggregator = self._get_unique_function_id(other._aggregator)
+
+        if(
+            self_select == other_select and
+            self_aggregator == other_aggregator
+        ):
+            return True
+
+        return False
+
+    def __hash__(self):
+        blob_l = list(map(hash, [self._sort_by, self._reverse_order]))
+        # Get unique id for _aggregator and _select attributes
+        blob_l += list(map(self._get_unique_function_id, [self._aggregator, self._select]))
+        blob = ','.join(str(attr) for attr in blob_l)
+
+        return int(sha1(blob.encode('utf-8')).hexdigest(), 16)
+
+    def _get_unique_function_id(self, func):
```

Comment:
I would explain what this function is doing as it is not immediately apparent. I understand, but someone fresh coming to the code might not.

Reply:
This is indeed better, but the main purpose it to provide a way to compare arbitrary types that are callable, and that isn't mentioned here.